examples/cc3200/xively_firmware_updates/SFT/xi_sft_client.c

```diff
@@ -0,0 +1,1336 @@
+/*  Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+    This is part of Xively C library. */
+
+
+/*
+ * HEADERS
+ */
+
+/* SFT Client Includes */
+#include <xively.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "../cbor/cbor.h"
+#include "../cbor/cn-cbor/cn-cbor.h"
+#include "../sha256/sha256.h"
+#include "xi_sft.h"
+
+/* Simpelink EXT lib includes */
+#include "flc_api.h"
+
+/* Platform includes */
+#include "hw_memmap.h"
+#include "hw_common_reg.h"
+#include "hw_types.h"
+#include "hw_ints.h"
+#include "interrupt.h"
+#include "rom.h"
+#include "rom_map.h"
+#include "uart.h"
+#include "prcm.h"
+#include "utils.h"
+#include "gpio.h"
+#include "gpio_if.h"
+#include "uart_if.h"
+
+/*
+ * DEFINES
+ */
+#define UART_PRINT Report
+#undef printf
+#define printf Report
+
+#define MAX_INCOMING_FILENAME_LENGTH 50
+#define MAX_STRING_LENGTH 50
+#define MESSAGESIZE 50
+#define LOGDETAILSSIZE 400
+
+/*
+ * TYPEDEFS
+ */
+typedef unsigned char uchar;
+typedef unsigned int uint;
+
+typedef enum {
+    sft_status_ok = 0, /* 0 */
+
+    sft_status_rebooting, /* 1 */
+
+    sft_open_file_error,  /* 2 */
+    sft_close_file_error, /* 3 */
+    sft_write_error,      /* 4 */
+
+    sft_unsupported_protocol_version, /* 5 */
+    sft_unexpected_filename_error,    /* 6 */
+    sft_empty_length_array_error,     /* 7 */
+    sft_data_out_of_bounds_error,     /* 8 */
+    sft_file_data_offset_error,       /* 9 */
+
+    sft_cbor_missing_field_error,   /* 10 */
+    sft_cbor_encoding_error,        /* 11 */
+    sft_cbor_error_field_set_error, /* 12 */
+
+    sft_publish_error,               /* 13 */
+    sft_internal_error,              /* 14 */
+    sft_fingerprint_mismatch_error,  /* 15 */
+    sft_fingerprint_too_large_error, /* 16 */
+
+    sft_status_max
+
+} sft_status_t;
+
+/* stores state data for the download state machine */
+typedef struct file_download_ctx_s
+{
+    char is_downloading;
+    int file_length;
+    int file_storage_offset;
+    sft_status_t result;
+
+    long file_handle;
+    unsigned long file_token;
+
+    char file_name[MAX_STRING_LENGTH];
+    char file_revision[MAX_STRING_LENGTH];
+    unsigned char file_sft_fingerprint_str[65];
+    unsigned char file_local_computed_fingerprint[32];
+
+    SHA256_CTX sha256_context;
+
+} file_download_ctx_t;
+
+/* data is extracted from FILE_CHUNK cbor messages and
+ * temporarily stored in this more accessible structure */
+typedef struct xi_sft_file_chunk_s
+{
+    uint offset;
+    uint chunk_length;
+    uint array_length;
+    uint status;
+    unsigned char* byte_array;
+} xi_sft_file_chunk_t;
+
+
+/*
+ * VARIABLES
+ */
+int tenpercent        = 0;
+int currenttenpercent = 0;
+
+char firmware_filename[] = "firmware";
+char firmware_revision[] = "1.0";
+
+file_download_ctx_t download_ctx;
+
+/* Buffers to hold the formatted topics.
+ * Topics are mangled by account and device ids.
+ */
+char xi_stopic[128];
+char xi_logtopic[128];
```

Comment:
These are static and not up to the customer to use for their own topic buffers. snprintf prevents overflow, and the svc , cln and _log channels fit well within this limitation.